File: src/test/java/application/cast/CastCreatureTest.java
package application.cast;

import application.AbstractApplicationTest;
import com.aa.mtg.MtgApplication;
import com.aa.mtg.game.init.test.InitTestService;
import com.aa.mtg.game.status.GameStatus;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit4.SpringRunner;

import static application.browser.BattlefieldHelper.FIRST_LINE;
import static application.browser.BattlefieldHelper.SECOND_LINE;
import static com.aa.mtg.cards.Cards.ISLAND;
import static com.aa.mtg.cards.Cards.MOUNTAIN;
import static com.aa.mtg.cards.sets.Ixalan.HEADWATER_SENTRIES;
import static com.aa.mtg.game.player.PlayerType.OPPONENT;
import static com.aa.mtg.game.player.PlayerType.PLAYER;
import static com.aa.mtg.game.turn.phases.Main1Phase.M1;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = MtgApplication.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Import({CastCreatureTest.InitGameTestConfiguration.class})
public class CastCreatureTest extends AbstractApplicationTest {
    @Test
    public void castCreature() {
        // When click on creature without paying the cost
        browser.player1().getHandHelper(PLAYER).getFirstCard(HEADWATER_SENTRIES).click();

        // Then stack is still empty
        browser.player1().getStackHelper().toHaveSize(0);

        // When clicking all lands
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 0).click();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 1).click();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 2).click();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 0).click();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 1).click();

        // Then all lands are front-end tapped
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 0).isFrontendTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 1).isFrontendTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 2).isFrontendTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 0).isFrontendTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 1).isFrontendTapped();

        // When click on creature
        browser.player1().getHandHelper(PLAYER).getCard(HEADWATER_SENTRIES, 0).click();

        // Then all lands are tapped for both players
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 0).isTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 1).isTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(ISLAND, 2).isTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 0).isTapped();
        browser.player1().getBattlefieldHelper(PLAYER, FIRST_LINE).getCard(MOUNTAIN, 1).isTapped();

        browser.player2().getBattlefieldHelper(OPPONENT, FIRST_LINE).getCard(ISLAND, 0).isTapped();
        browser.player2().getBattlefieldHelper(OPPONENT, FIRST_LINE).getCard(ISLAND, 1).isTapped();
        browser.player2().getBattlefieldHelper(OPPONENT, FIRST_LINE).getCard(ISLAND, 2).isTapped();
        browser.player2().getBattlefieldHelper(OPPONENT, FIRST_LINE).getCard(MOUNTAIN, 0).isTapped();
        browser.player2().getBattlefieldHelper(OPPONENT, FIRST_LINE).getCard(MOUNTAIN, 1).isTapped();

        // And creature is on the stack for both players (hands is empty)
        browser.player1().getStackHelper().containsExactly(HEADWATER_SENTRIES);
        browser.player1().getHandHelper(PLAYER).isEmpty();
        browser.player2().getStackHelper().containsExactly(HEADWATER_SENTRIES);
        browser.player1().getHandHelper(OPPONENT).isEmpty();

        // And priority is to opponent with corresponding status
        browser.player1().getPhaseHelper().is(M1, OPPONENT);
        browser.player1().getStatusHelper().hasMessage("Wait for opponent to perform its action...");
        browser.player1().getActionHelper().cannotContinue();
        browser.player2().getPhaseHelper().is(M1, PLAYER);
        browser.player2().getStatusHelper().hasMessage("Play any instant or ability or resolve the top spell in the stack (SPACE).");
        browser.player2().getActionHelper().canContinue();

        // When player2 acknowledge the spell casted
        browser.player2().getActionHelper().clickContinue();

        // Creature goes on the battlefield (stack is empty)
        browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).contains(HEADWATER_SENTRIES);
        browser.player2().getBattlefieldHelper(OPPONENT, SECOND_LINE).contains(HEADWATER_SENTRIES);
        browser.player1().getStackHelper().isEmpty();
        browser.player2().getStackHelper().isEmpty();

        // And priority is to player again
        browser.player1().getPhaseHelper().is(M1, PLAYER);
        browser.player1().getStatusHelper().hasMessage("Play any spell or ability or continue (SPACE).");
        browser.player1().getActionHelper().canContinue();
        browser.player2().getPhaseHelper().is(M1, OPPONENT);
        browser.player2().getStatusHelper().hasMessage("Wait for opponent to perform its action...");
        browser.player2().getActionHelper().cannotContinue();

        // Hand is now empty
        browser.player1().getHandHelper(PLAYER).isEmpty();
        browser.player2().getHandHelper(PLAYER).isEmpty();
    }

    @Configuration
    static class InitGameTestConfiguration {
        @Bean
        public InitTestService initTestService() {
            return new InitTestService() {
                @Override
                protected void initGameStatus(GameStatus gameStatus) {
                    addCardToCurrentPlayerHand(gameStatus, HEADWATER_SENTRIES);
                    addCardToCurrentPlayerBattlefield(gameStatus, ISLAND);
                    addCardToCurrentPlayerBattlefield(gameStatus, ISLAND);
                    addCardToCurrentPlayerBattlefield(gameStatus, ISLAND);
                    addCardToCurrentPlayerBattlefield(gameStatus, MOUNTAIN);
                    addCardToCurrentPlayerBattlefield(gameStatus, MOUNTAIN);
                }
            };
        }
    }
}

File: src/test/kotlin/application/cast/CastCreatureTest.kt
package application.cast

import application.AbstractApplicationTest
import application.InitTestService
import application.browser.BattlefieldHelper
import com.matag.cards.Cards
import com.matag.cards.properties.Color
import com.matag.game.cardinstance.CardInstanceFactory
import com.matag.game.status.GameStatus
import com.matag.game.turn.phases.main1.Main1Phase
import com.matag.player.PlayerType
import org.junit.jupiter.api.Test
import org.springframework.context.annotation.Import

@Import(CastCreatureTest.InitTestServiceForTest::class)
class CastCreatureTest(var initService: InitTestService) : AbstractApplicationTest() {

    override fun setupGame() {
        initService = InitTestServiceForTest(cardInstanceFactory, cards)
    }

    @Test
    fun castCreature() {
        // When click on creature without paying the cost
        browser.player1().getHandHelper(PlayerType.PLAYER).getFirstCard(cards.get("Headwater Sentries")).click()

        // Then stack is still empty
        browser.player1().getStackHelper().toHaveSize(0)

        // When clicking all lands
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 0).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 1).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 2).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 1).click()

        // Then all lands are front-end tapped
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 0).isFrontendTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 1).isFrontendTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 2).isFrontendTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).isFrontendTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 1).isFrontendTapped()
        browser.player1().getPlayerActiveManaHelper().toHaveMana(
            listOf(
                Color.BLUE,
                Color.BLUE,
                Color.BLUE,
                Color.RED,
                Color.RED
            )
        )

        // When clicking on a land again then gets untapped and when clicking on it again then gets tapped again
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).isNotFrontendTapped()
        browser.player1().getPlayerActiveManaHelper().toHaveMana(
            listOf(
                Color.BLUE,
                Color.BLUE,
                Color.BLUE,
                Color.RED
            )
        )
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).click()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).isFrontendTapped()
        browser.player1().getPlayerActiveManaHelper().toHaveMana(
            listOf(
                Color.BLUE,
                Color.BLUE,
                Color.BLUE,
                Color.RED,
                Color.RED
            )
        )

        // When click on creature
        browser.player1().getHandHelper(PlayerType.PLAYER).getCard(cards.get("Headwater Sentries"), 0).click()

        // Then all lands are tapped for both players
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 0).isTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 1).isTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 2).isTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).isTapped()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 1).isTapped()

        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 0).isTapped()
        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 1).isTapped()
        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Island"), 2).isTapped()
        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 0).isTapped()
        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.FIRST_LINE)
            .getCard(cards.get("Mountain"), 1).isTapped()

        // And creature is on the stack for both players (hands is empty)
        browser.player1().getStackHelper().containsExactly(cards.get("Headwater Sentries"))
        browser.player1().getHandHelper(PlayerType.PLAYER).isEmpty
        browser.player2().getStackHelper().containsExactly(cards.get("Headwater Sentries"))
        browser.player1().getHandHelper(PlayerType.OPPONENT).isEmpty

        // And priority is to opponent with corresponding status
        browser.player1().getPhaseHelper().matches(Main1Phase.M1, PlayerType.OPPONENT)
        browser.player1().getStatusHelper().hasMessage("Wait for opponent to perform its action...")
        browser.player1().getActionHelper().cannotContinue()
        browser.player2().getPhaseHelper().matches(Main1Phase.M1, PlayerType.PLAYER)
        browser.player2().getStatusHelper().hasMessage("Play any instant or ability or resolve the top spell in the stack (SPACE).")
        browser.player2().getActionHelper().canContinue()

        // When player2 acknowledge the spell casted
        browser.player2().getActionHelper().clickContinueAndExpectPhase(Main1Phase.M1, PlayerType.PLAYER)

        // Creature goes on the battlefield (stack is empty)
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .contains(cards.get("Headwater Sentries"))
        browser.player2().getBattlefieldHelper(PlayerType.OPPONENT, BattlefieldHelper.SECOND_LINE)
            .contains(cards.get("Headwater Sentries"))
        browser.player1().getStackHelper().isEmpty
        browser.player2().getStackHelper().isEmpty

        // And priority is to player again
        browser.player1().getPhaseHelper().matches(Main1Phase.M1, PlayerType.PLAYER)
        browser.player1().getStatusHelper().hasMessage("Play any spell or ability or continue (SPACE).")
        browser.player1().getActionHelper().canContinue()
        browser.player2().getPhaseHelper().matches(Main1Phase.M1, PlayerType.OPPONENT)
        browser.player2().getStatusHelper().hasMessage("Wait for opponent to perform its action...")
        browser.player2().getActionHelper().cannotContinue()

        // Hand is now empty
        browser.player1().getHandHelper(PlayerType.PLAYER).isEmpty
        browser.player2().getHandHelper(PlayerType.PLAYER).isEmpty
    }

    internal class InitTestServiceForTest(cardInstanceFactory: CardInstanceFactory, cards: Cards) : InitTestService(cardInstanceFactory, cards) {
        override fun initGameStatus(gameStatus: GameStatus) {
            addCardToCurrentPlayerHand(gameStatus, cards.get("Headwater Sentries"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Island"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Island"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Island"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Mountain"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Mountain"))
        }
    }
}
